### PR TITLE
GH-36240: [Python] Refactor CumulativeSumOptions to a separate class for independent deprecation

### LINKED ISSUE #36240: [Python] Refactor CumulativeSumOptions to a separate class for independent deprecation
### Describe the enhancement requested

CumulativeSumOptions are renamed to CumulativeOptions in https://github.com/apache/arrow/pull/36020 after the introduction of cumulative functions other than `cumulative_sum` which reuse those options. For backward compatibility, [an alias](https://github.com/apache/arrow/pull/36020/commits/c61d6cf2df5a7c83a8aa4cbf51ed01c5d18165b6#diff-3d08757408024228c4443730cc3536ab39c9436cd2e4cb63e5da34c69c18962fR37) was added so that the name CumulativeSumOptions is still usable. However [per discussion on that PR](https://github.com/apache/arrow/pull/36020#discussion_r1238433064) the old name should be deprecated, which requires an explicit separate class.

### Component(s)

Python

### PR DESCRIPTION
**Pull request by R-JunmingChen**
**Rationale for this change**
As https://github.com/apache/arrow/issues/36240 says, we refactor CumulativeSumOptions to a separate class.

**What changes are included in this PR?**

- independent CumulativeSumOptions 
- the original simple test before #32190 
- fix a typo in CumulativeOptions 

**Are these changes tested?**
No.
Actually, the PR can't pass the `test_option_class_equality` in test_compute.py ([Error example](https://github.com/apache/arrow/actions/runs/5728571658/job/15523443371?pr=36977)). Cause CumulativeSumOptions's C++ part is also CumulativeOptions.
image

**Are there any user-facing changes?**
No.

Closes: https://github.com/apache/arrow/issues/36240
* Closes: #36240

### CODE REVIEW

**Comment by R-JunmingChen** [comment]
Hi, reviewers,here are two choices to handle the problem the PR can't pass `test_option_class_equality`.  Do you have any ideas about it?

- We ignore it, this unequality seems to have no impact on usage for users.
- We also make CumulativeSumOptions a independent class in C++.

File: python/pyarrow/_compute.pyx [hunk 1]
```diff
@@ -1933,6 +1933,34 @@ class PartitionNthOptions(_PartitionNthOptions):
...
+    ----------
+    start : Scalar, default 0.0
+        Starting value for sum computation
+    skip_nulls : bool, default False
+        When false, the first encountered null is propagated.
+    """
+
+    def __init__(self, start=0.0, *, skip_nulls=False):
+        self._set_options(start, skip_nulls)
+
```
Comment (danepitkin) [inline python/pyarrow/_compute.pyx:1962]: We can go ahead and add the deprecation warning with https://github.com/apache/arrow/blob/f44e28fa03a64ae5b3d9352d21aee2cc84f9af6c/python/pyarrow/util.py#L109

File: python/pyarrow/_compute.pyx [hunk 2]
```diff
@@ -1933,6 +1933,34 @@ class PartitionNthOptions(_PartitionNthOptions):
...
+    Parameters
+    ----------
+    start : Scalar, default 0.0
+        Starting value for sum computation
+    skip_nulls : bool, default False
+        When false, the first encountered null is propagated.
+    """
+
+    def __init__(self, start=0.0, *, skip_nulls=False):
```
Comment (danepitkin) [inline python/pyarrow/_compute.pyx:1960]: We should keep the implementation identical to `CumulativeOptions`, which uses `start=None`. Arrow v13 will release with the new `CumulativeOptions`, but this PR will release in Arrow v14. So the `start=None` change will already be made.

File: python/pyarrow/_compute.pyx [hunk 3]
```diff
@@ -1947,7 +1975,7 @@ cdef class _CumulativeOptions(FunctionOptions):
                     pyarrow_unwrap_scalar(start), skip_nulls))
             except Exception:
                 _raise_invalid_function_option(
-                    start, "`start` type for CumulativeSumOptions", TypeError)
+                    start, "`start` type for CumulativeOptions", TypeError)
```
Comment (danepitkin) [inline python/pyarrow/_compute.pyx:1978]: Nice catch!

File: python/pyarrow/_compute.pyx [hunk 4]
```diff
@@ -1933,6 +1933,34 @@ class PartitionNthOptions(_PartitionNthOptions):
         self._set_options(pivot, null_placement)
 
 
+cdef class _CumulativeSumOptions(FunctionOptions):
```
Comment (danepitkin) [inline python/pyarrow/_compute.pyx:1936]: Could we delete this cdef class and just inherit from `_CumulativeOptions`?

**Review by danepitkin — COMMENTED** [review]
For the failing test, we can just add an explicit check for CumulativeSumOptions == CumulativeOptions. We can add a TODO comment to remove the check when the deprecated class is removed.

File: python/pyarrow/_compute.pyx (hunk 4 above)
Comment (R-JunmingChen) [reply python/pyarrow/_compute.pyx:1936]: Sure, Have deleted. 
But if we do that, the error message will report  `"`start` type for CumulativeOptions"` for CumulativeSumOptions.

File: python/pyarrow/_compute.pyx (hunk 1 above)
Comment (R-JunmingChen) [reply python/pyarrow/_compute.pyx]: Hi, I am not sure how to use `_deprecate_class`, which doesn't look like a decorator.  It seems that we can use like this:

`CumulativeSumOptions = _deprecate_class("CumulativeSumOptions", CumulativeOptions, 14)`

which can't be referenced when using CumulativeSumOptions with pyarrow.compute CumulativeSumOptions.

Comment (R-JunmingChen) [reply python/pyarrow/_compute.pyx]: May be we use alias

`Deprecate_CumulativeSumOptions = _deprecate_class("CumulativeSumOptions", CumulativeOptions, 14)` in _compute.py

`Deprecate_CumulativeSumOptions as  CumulativeSumOptions,` in compute.py

File: python/pyarrow/_compute.pyx (hunk 4 above)
Comment (danepitkin) [reply python/pyarrow/_compute.pyx]: Ah good point. I think that's okay, but I am also okay with duplicating it if you prefer that instead.

File: python/pyarrow/_compute.pyx (hunk 1 above)
Comment (danepitkin) [reply python/pyarrow/_compute.pyx]: Hmm, maybe `_deprecate_api` should be used instead then. That does provide a wrapper around the API. If that doesn't work for this use case either, than just a manual warning in the `__init__` method will suffice.

Comment (R-JunmingChen) [reply python/pyarrow/_compute.pyx]: I add a manual warning

File: python/pyarrow/_compute.pyx (hunk 4 above)
Comment (R-JunmingChen) [reply python/pyarrow/_compute.pyx]: Ok. Let's keep the deteled version

File: python/pyarrow/_compute.pyx [hunk 5]
```diff
@@ -1974,6 +1975,26 @@ class CumulativeOptions(_CumulativeOptions):
...
+        Starting value for sum computation
+    skip_nulls : bool, default False
+        When false, the first encountered null is propagated.
+    """
+
+    def __init__(self, start=None, *, skip_nulls=False):
+        warnings.warn(
+            _DEPR_MSG.format("CumulativeSumOptions", "14.0", "CumulativeOptions"),
+            FutureWarning,
+        )
```
Comment (danepitkin) [inline python/pyarrow/_compute.pyx:1994]:
```suggestion
            FutureWarning,
            stacklevel=2
        )
```

**Comment by danepitkin** [comment]
One suggestion, otherwise LGTM! Nice work! I'll add an Arrow committer to the reviewer's list.

File: python/pyarrow/tests/test_compute.py [hunk 6]
```diff
@@ -210,7 +210,12 @@ def test_option_class_equality():
         buf = option.serialize()
         deserialized = pc.FunctionOptions.deserialize(buf)
         assert option == deserialized
-        assert repr(option) == repr(deserialized)
+        # TODO remove the check under if
+        # when the deprecated class CumulativeSumOptions is removed.
```
Comment (AlenkaF) [inline python/pyarrow/tests/test_compute.py:214]:
```suggestion
        # TODO remove the check under the if statement
        # when the deprecated class CumulativeSumOptions is removed.
```

**Review by AlenkaF — APPROVED** [review]
LGTM +1

Just a small nitpick suggestion for clearer understanding of the TODO, then I will merge 👍

**Pull request merged**